### DISCUSSION TITLE
New API Endpoint "/mystar-v2"

### DISCUSSION THREAD

**Discussion by loebse** [pypolestar/polestar_api#95] [General]
Looks like there is a new API endpoint available. 

The "old" ``/my-star`` is no longer working. However, there is a new one ``/mystar-v2``

**Comment by loebse** [comment]
Are you aware of this already @leeyuentuen ?

**Comment by leeyuentuen** [comment]
let me check it

**Reply by Ottmar0815** [reply]
Whats the latest beta witch fixes the new endpoint?

**Comment by leeyuentuen** [comment]
public release 1.6.0

**Comment by andig** [comment]
Developer here, coming from https://github.com/leeyuentuen/polestar_api/issues/96#issuecomment-1905553439:

> i haven't use in this api the getAuthToken. but the graphql should be like this:

Not sure I follow- I would save you are using it?

```python
    if not refresh or self.token_expiry is None or self.token_expiry < datetime.now():
            code = await self._get_code()
            if code is None:
                return
            params = {
                "query": "query getAuthToken($code: String!) { getAuthToken(code: $code) { id_token access_token refresh_token expires_in }}",
                "operationName": operationName,
                "variables": json.dumps({"code": code}),
            }
```

Trying this I always get failure:

```
POST /eu-north-1/auth HTTP/1.1
Host: pc-api.polestar.com
User-Agent: Go-http-client/1.1
Content-Length: 216
Content-Type: application/json
Accept-Encoding: gzip

{"query":"query getAuthToken($code:String!){getAuthToken(code: $code){id_token,access_token,refresh_token,expires_in}}","variables":{"code":"bx..."},"operationName":"getAuthToken"}

{"errors":[{"message":"An internal error has occured.","path":["getAuthToken"],"extensions":{"code":"BAD_USER_INPUT","exception":{"stacktrace":["UserInputError: An internal error has occured.","    at ULi (/var/task/src/functions/auth/handler.js:1322:2696188)","    at runMicrotasks (<anonymous>)","    at processTicksAndRejections (node:internal/process/task_queues:96:5)","    at async GLi (/var/task/src/functions/auth/handler.js:1322:2697017)","    at async getAuthToken (/var/task/src/functions/auth/handler.js:1322:2723641)"],"message":"An internal error has occured.","locations":[{"line":1,"column":44}],"path":["getAuthToken"]},"serviceName":"auth"}}],"data":{"getAuthToken":null}}
```

Changing `operationName` to something uppercase doesn't make a difference either. I'm totally lost whats wrong.

> by the way, it would be interesting if you were using polexplore. on my API version here I'm using polmystar
> if you have token from polexplore, then you should also get access on the climate and the door unlock feature I suppose

I was using `polexplore`: https://github.com/evcc-io/evcc/blob/3e23dd12e908ee14faf814e3e66c513f247695fd/vehicle/polestar/identity.go#L27 never tried more than soc/odo.

**Reply by leeyuentuen** [reply]
i'll take a look later

**Reply by loebse** [reply]
I tried to create a small example written in Go, I guess this is what you need @andig. You can have a look here: https://github.com/loebse/polestar_go_example/blob/main/polestarAuth.go

However, I am not really familiar with Go and get stuck at step 2 in the following process that I think we must go through:

1.  Get resumePath -> this works fine in my example
2. Take the resume path and getCode -> here I am either getting a code ``200`` when I am following the redirect or a code ``303`` when I am not following. However, I guess we need a code ``302`` here looking at your code @leeyuentuen Maybe you have an idea what must change to make this work @andig ?
3. Now we would need to get the Token but I never get here because of step 2

Maybe this helps in your investigation @andig ?

**Reply by leeyuentuen** [reply]
302 is used in http2 to know to redirect to the next page. if you get a 200, the content of the body should have an issue.

i didn't look in detail, but it could be that the polxplore client_id and client_secret has been changed in that code
the basic auth I saw is:
cG9seHBsb3JlOlhhaUtvb0hlaXJlaXNvb3NhaDBFdjZxdW9oczhjb2hGZUtvaHdpZTFhZTdraWV3b2hkb295ZWk5QWVZZWlXb2g

that is client_id: polxplore
client_secret: XaiKooHeireisoosah0Ev6quohs8cohFeKohwie1ae7kiewohdooyei9AeYeiWoh

maybe that client_Secret changed

**Reply by andig** [reply]
Finally fixed. I was still sending pkce challenge during code exchange which results in exchange broken when retrieving token by code. Thanks all for supporting!
